service/history/activityReplicator.go

```diff
@@ -0,0 +1,242 @@
+	shard ShardContext,
+	historyCache *historyCache,
+	logger log.Logger,
+) *activityReplicatorImpl {
+
+	return &activityReplicatorImpl{
+		historyCache:    historyCache,
+		clusterMetadata: shard.GetService().GetClusterMetadata(),
+		logger:          logger.WithTags(tag.ComponentHistoryReplicator),
+	}
+}
+
+func (r *activityReplicatorImpl) SyncActivity(
+	ctx ctx.Context,
+	request *h.SyncActivityRequest,
+) (retError error) {
+
+	// sync activity info will only be sent from active side, when
+	// 1. activity has retry policy and activity got started
+	// 2. activity heart beat
+	// no sync activity task will be sent when active side fail / timeout activity,
+	// since standby side does not have activity retry timer
+
+	domainID := request.GetDomainId()
+	execution := workflow.WorkflowExecution{
+		WorkflowId: request.WorkflowId,
+		RunId:      request.RunId,
+	}
+
+	context, release, err := r.historyCache.getOrCreateWorkflowExecution(ctx, domainID, execution)
+	if err != nil {
+		// for get workflow execution context, with valid run id
+		// err will not be of type EntityNotExistsError
+		return err
+	}
+	defer func() { release(retError) }()
+
+	msBuilder, err := context.loadWorkflowExecution()
+	if err != nil {
+		if _, ok := err.(*workflow.EntityNotExistsError); !ok {
+			return err
+		}
+
+		// this can happen if the workflow start event and this sync activity task are out of order
+		// or the target workflow is long gone
+		// the safe solution to this is to throw away the sync activity task
+		// or otherwise, worker attempt will exceeds limit and put this message to DLQ
+		return nil
+	}
+
+	version := request.GetVersion()
+	scheduleID := request.GetScheduledId()
+	if msBuilder.GetVersionHistories() != nil {
+		currentVersionHistory, err := msBuilder.GetVersionHistories().GetCurrentVersionHistory()
+		if err != nil {
+			return err
+		}
+		lastLocalItem, err := currentVersionHistory.GetLastItem()
+		if err != nil {
+			return err
+		}
+		incomingVersionHistory := persistence.NewVersionHistoryFromThrift(request.GetVersionHistory())
+		lastIncomingItem, err := incomingVersionHistory.GetLastItem()
+		if err != nil {
+			return err
+		}
+		if lastIncomingItem.GetVersion() < lastLocalItem.GetVersion() {
+			// the incoming branch will lose to the local branch
+			// discard this task
+			return nil
+		}
+
+		lcaItem, err := currentVersionHistory.FindLCAItem(incomingVersionHistory)
+		if err != nil {
+			return err
+		}
+
+		if !currentVersionHistory.IsLCAAppendable(lcaItem) || scheduleID >= lastLocalItem.GetVersion() {
+			// incoming branch will win the local branch
+			// resend the events with higher version
+			return newNDCRetryTaskErrorWithHint(
+				domainID,
+				execution.GetWorkflowId(),
+				execution.GetRunId(),
+				common.Int64Ptr(lcaItem.GetEventID()),
+				common.Int64Ptr(lcaItem.GetVersion()),
+				common.Int64Ptr(lastIncomingItem.GetEventID()),
+				common.Int64Ptr(lastIncomingItem.GetVersion()),
```

Comment:
this maybe an overkill, i.e. sending all events until lastIncomingItem
is there anyway to optimize this?

Reply:
Added a optimization